QccPack-0.61-1/src/libQccPackSQ/utils/man1/sqlloyd.1
.TH SQLLOYD 1 "QCCPACK" ""
.SH NAME
sqlloyd \- design Lloyd optimal mean-square scalar quantizer for Gaussian and 
Laplacian densities
.SH SYNOPSIS
.B sqlloyd
.RB "[\|" \-max
.IR  max\_value "\|]"
.RB "[\|" \-min
.IR min\_value "\|]"
.RB "[\|" \-nl
.IR num\_levels "\|]"
.RB "[\|" \-v
.IR variance "\|]"
.RB "[\|" \-l "\|]"
.RB "[\|" \-t
.IR stop\_threshold "\|]"
.RB "[\|" \-i
.IR integration\_intervals "\|]"
.I quantizer\_file
.SH OPTIONS
.TP
.BI \-max " max\_value"
Float. Maximum value of the range of the quantizer.  Default value = 10.
.TP
.BI \-min " min\_value"
Float. Minimum value of the range of the quantizer.  Default value = -10.
.TP
.BI \-nl " num\_levels"
Integer.  Number of quantizer intervals. Default = 255.
.TP
.BI \-v " variance"
Float.  Variance of the Gaussian or Laplacian density.  Default = 1.0.
.TP
.B \-l
Use Laplacian density.  The default is to use a Gaussian density.
.TP
.BI \-t " stop\_threshold"
Float.  Stop the design process when the average distortion (MSE) 
changes by less than 
.I stop\_threshold
from one iteration to the next; that is, stop if
(D(i-1) - D(i))/D(i) <
.IR threshold ,
where D(i) is the average distortion (MSE) for the ith iteration.
.TP
.BI \-i " integration\_intervals"
Float.  Specifies the number of intervals to use in the numeric integration
used tp calculate partition probabilities and centroids.
.SH DESCRIPTION
.LP
.B sqlloyd
designs a scalar quantizer to minimize the mean-square error (MSE)
for the given probability density and number of quantizer levels.
The designed quantizer is output to
.IR quantizer\_file .
The range of the quantizer is 
.I min\_value 
through
.IR max\_value .
The number of quantization levels is
.IR num\_levels .
.LP
If the
.B \-l
option is specified, a Laplacian density is used.  Otherwise, a Gaussian
density is used.  In either case, 
.I variance
is the variance of the density to be used.
The mean of the density is assumed to be
.RI ( max\_value + min\_value ")/2."
.LP
The Lloyd algorithm involves finding the first moment of each partition
and the probability of each partition.  Numerical integration using
Simpon's Rule is used in each of these calculations.  The value
of
.I integration\_intervals
gives the number of intervals to use in Simpson's Rule.
The more intervals used, the more accurate will be the the integration, but
using more levels increases the computation time.  100 to 1000 intervals
is probably sufficient for most applications.
.LP
The design process stops when the average distortion of the quantizer changes
by less than
.I stop\_threshold
between iterations.  Small values yield better quantizers yet longer
computation time.  Values in the range 0.001 to 0.000001 are probably 
sufficient for most applications.
.SH "SEE ALSO"
.BR sqlloyd (1),
.BR QccPackSQ (3),
.BR QccPack (3)

A. Gersho and R. Gray, 
.IR "Vector Quantization and Signal Compression" .
Norwell, MA: Kluwer Academic Publishers, 1992, pp. 187-194.

A. K. Jain,
.IR "Fundamentals of Digital Image Processing" .
Englewood Cliffs, NJ: Prentice Hall, 1989, pp. 101-112.

.SH AUTHOR
Copyright (C) 1997-2016  James E. Fowler
.\"  The programs herein are free software; you can redistribute them and/or
.\"  modify them under the terms of the GNU General Public License
.\"  as published by the Free Software Foundation; either version 2
.\"  of the License, or (at your option) any later version.
.\"  
.\"  These programs are distributed in the hope that they will be useful,
.\"  but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"  GNU General Public License for more details.
.\"  
.\"  You should have received a copy of the GNU General Public License
.\"  along with these programs; if not, write to the Free Software
.\"  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.


QccPack-0.61-1/CHANGELOG
==============================================================================

                                  Q C C P A C K

                            Version 0.61, 01-apr-2016

                    Copyright (C) 1997-2016  James E. Fowler

------------------------------------------------------------------------------


CHANGES:

Version 0.61:
  01-apr-2016:  BETA release version 0.61
  30-mar-2016:  Updated Cygwin installation instructions
  20-jan-2016:  Updated copyright year range for 2016
  20-jan-2016:  Fixed zero-subband-size bug in tce3d.c

Version 0.60:
  25-nov-2015:  BETA release version 0.60
  08-aug-2015:  Added QccFileReadLine()
  07-aug-2015:  Added QccFileGetRealPath()
  14-jul-2015:  Added QccFileWriteMagicNumberVersion()
  
Version 0.59:
  18-jun-2015:  BETA release version 0.59
  15-jun-2015:  Added QccFileRewind()
  14-jun-2015:  Updated copyright year range for 2015
  14-jun-2015:  Made QccENTExponentialGolombEncodeSymbol() and
                  QccENTExponentialGolombDecodeSymbol() visible
  09-jan-2014:  Updated copyright year range for 2014
  20-jul-2013:  Added Data/Matlab to QCCPACK_DATA_DIRS in Imakefile
  20-jul-2013:  Updated copyright year range for 2013
  21-dec-2011:  Updated copyright year range for 2011
  21-dec-2011:  Minor fixes for warning-free compilation with gcc-4.6
  23-jun-2009:  Added better support for uniform quantizers to sqencode,
                  sqdecode
  11-mar-2009:  Added logo.sed

Version 0.58:
  13-feb-2009:  BETA release version 0.58
  05-feb-2009:  Added klttcedecode3d.1, klttcedecode3d_lossless.1, 
                  klttceencode3d.1, klttceencode3d_lossless.1, 
                  tcedecode3d_lossless.1, tceencode3d_lossless.1
  05-feb-2009:  Added QccWAVtce3DLosslessEncode.3,
                  QccWAVtce3DLosslessDecodeHeader.3, 
                  QccWAVtce3DLosslessDecode.3, 
                  QccWAVklttce3DEncode.3, 
                  QccWAVklttce3DDecodeHeader.3, 
                  QccWAVklttce3DDecode.3, 
                  QccWAVklttce3DLosslessEncode.3, 
                  QccWAVklttce3DLosslessDecodeHeader.3, 
                  QccWAVklttce3DLosslessDecode.3
  05-feb-2009:  Added klttce3d.c, klttceencode3d.c, klttcedecode3d.c
  04-feb-2009:  Changed QccHYPrkltTrain(), QccHYPrkltTransform(),
                  QccHYPrkltInverseTransform() to use integer volumes
  04-feb-2009:  Added klttce3d_lossless.c, klttceencode3d_lossless.c,
                  klttcedecode3d_lossless.c
  04-feb-2009:  Fixed zero temporal_num_scales problem in 3D packet transforms
  02-feb-2009:  Added tce3d_lossless.c,
                  tceencode3d_lossless, tcedecode3d_lossless
  29-jan-2009:  QccHYPrkltTrain.3, QccHYPrkltFactorization.3,
                  QccHYPrkltTransform.3, QccHYPrkltInverseTransform.3
  29-jan-2009:  Added QccHYPrklt.3
  29-jan-2009:  Added QccHYPrkltInitialize(), QccHYPrkltAlloc(),
                  QccHYPrkltFree(), QccHYPrkltTrain(),
                  QccHYPrkltFactorization()
  28-jan-2009:  Added QccHYPrkltTransform(), QccHYPrkltInverseTransform()
  28-jan-2009:  Added QccMatrixIdentity(), QccMatrixRowExchange(),
                  QccMatrixColExchange(), QccMatrixIdentity.3,
                  QccMatrixRowExchange.3, QccMatrixColExchange.3
  16-jan-2009:  Updated copyright year range for 2009

Version 0.57:
  30-jun-2008:  BETA release version 0.57
  09-may-2008:  Added QccTimeTic(), QccTimeToc(), QccTime.3
  06-may-2008:  Fixed spatial_num_levels/temporal_num_levels bug in
                  initialization in bisk3d.c, wdr3d.c
  06-may-2008:  Fixed zero number of levels in
                  QccWAVWaveletPacketDWT3D(),
                  QccWAVWaveletInversePacketDWT3D()
  26-apr-2008:  Added sam.c, icbsam.c, QccHYPImageCubeMeanSAM(),
                QccVectorAngle()
  17-apr-2008:  Added QccWAVtce3DEncode.3, tceencode3d.1
  16-apr-2008:  Added tce3d.c, tceencode3d.c, tcedecode3d.c
  16-apr-2008:  Added -bpv to rawtoicb.c, icbtoraw.c, rawdist3d.c
  16-apr-2008:  Added -bpp to rawtoicp.c, icptoraw.c, rawdist2d.c
  16-apr-2008:  Added bpv, bpp to QccHYPRaw(), QccHYPRawRead2D(),
                  QccHYPRawWrite2D(), QccHYPRawDist2D(), QccHYPRawRead3D(),
                  QccHYPRawWrite3D(), QccHYPRawDist3D(),

Version 0.56:
  14-mar-2008:  BETA release version 0.56
  14-mar-2008:  Added QccIMGImageComponentLBT.3,
                  QccIMGImageComponentInverseLBT.3
  13-mar-2008:  Added QccFastDCT, QccFastDCTInitialize, QccFastDCTCreate,
                  QccFastDCTFree, QccFastDCTForwardTransform1D,
                  QccFastDCTInverseTransform1D, QccFastDCTForwardTransform2D,
                  QccFastDCTInverseTransform2D
  13-mar-2008:  Added fast_dct.c
  13-mar-2008:  Added dcttceencode.1, dcttcedecode.1
  13-mar-2008:  Added QccWAVdcttceEncode.3, QccWAVdcttceDecodeHeader.3,
                  QccWAVdcttceDecode.3
  13-mar-2008:  Renamed image_lapped_filter.c to image_lbt.c
  24-feb-2008:  Fixed memory leaks in dcttce.c, image_lapped_filter.c
  17-jan-2008:  Added Haar.fbk
  09-jan-2008:  Added dcttce.c, dcttceencode.c, dcttcedecode.c
  09-jan-2008:  Added image_lapped_filter.c
  09-jan-2008:  Added QccMatrixInverse(), QccMatrixInverse.3
  04-jan-2008:  Updated substr
  04-jan-2008:  Updated copyright year range for 2008
  11-dec-2007:  Fixed minor max_coefficients bug in QccWAVtceEncodeDWT()
  24-oct-2007:  Changed spawn to be able to handle commands with arguments
  09-oct-2007:  Added QCCMATHEPS
  01-oct-2007:  Added QccMatrixIntMultiply(), QccMatrixIntMultiply.3
  01-oct-2007:  Added QccMatrixMultiply(), QccMatrixMultiply.3
  28-sep-2007:  Added QccMatrixNullspace(), QccMatrixNullspace.3
  28-sep-2007:  Added QccMatrixOrthogonalize(), QccMatrixOrthogonalize.3
  28-sep-2007:  Modified QccMatrixSVD() to work with matrics of any size

Version 0.55:
  29-aug-2007:  BETA release version 0.55
  03-jul-2007:  Added limits.h include
  17-may-2007:  Fixed num_symbols bug in QccENTHuffmanDesign()

QccPack-0.61-1/src/libQccPackSQ/utils/sqlloyd.c
/*
 *
 * QccPack: Quantization, compression, and coding utilities
 * Copyright (C) 1997-2016  James E. Fowler
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "sqlloyd.h"

#define USG_STRING "[-max %f:max_value] [-min %f:min_value] [-nl %d:num_levels] [-v %f:variance] [-l %:] [-t %f:stop_threshold] [-i %d:integration_intervals] %s:quantizer_file"

QccSQScalarQuantizer Quantizer;

float MaxValue = 10;
float MinValue = -10;

int NumLevels = 255;

float Variance = 1.0;

float Threshold = 0.001;
int NumIntegrationIntervals = 100;

int Laplace = 0;

double (*ProbDensity)();

int main(int argc, char *argv[])
{

  QccInit(argc, argv);

  QccSQScalarQuantizerInitialize(&Quantizer);

  if (QccParseParameters(argc, argv,
                         USG_STRING,
                         &MaxValue,
                         &MinValue,
                         &NumLevels,
                         &Variance,
                         &Laplace,
                         &Threshold,
                         &NumIntegrationIntervals,
                         Quantizer.filename))
    QccErrorExit();
      
  if (NumLevels < 1)
    {
      QccErrorAddMessage("%s: Number of levels must be 1 or greater",
                         argv[0]);
      QccErrorExit();
    }
  Quantizer.num_levels = NumLevels;
  
  if (MaxValue <= MinValue)
    {
      QccErrorAddMessage("%s: MaxValue must be greater than MinValue",
                         argv[0]);
      QccErrorExit();
    }

  if (NumIntegrationIntervals < 10)
    {
      QccErrorAddMessage("%s: NumIntegrationIntervals must be 10 or greater",
                         argv[0]);
      QccErrorExit();
    }

  if (Laplace)
    ProbDensity = QccMathLaplacianDensity;
  else
    ProbDensity = QccMathGaussianDensity;

  if (QccSQLloydMakeQuantizer(&Quantizer,
                              ProbDensity,
                              (double)Variance,
                              (double)(MaxValue + MinValue)/2,
                              (double)MaxValue,
                              (double)MinValue,
                              (double)Threshold,
                              NumIntegrationIntervals))
    {
      QccErrorAddMessage("%s: Error calling QccSQLloydMakeQuantizer()",
                         argv[0]);
      QccErrorExit();
    }

  if (QccSQScalarQuantizerWrite(&Quantizer))
    {
      QccErrorAddMessage("%s: Error calling QccSQScalarQuantizerWrite()",
                         argv[0]);
      QccErrorExit();
    }

  QccSQScalarQuantizerFree(&Quantizer);

  QccExit;
}
